makefiles/features_provided_by_modules.inc.mk

```diff
@@ -0,0 +1,9 @@
+# Add features provided by modules
+#
+# This is done *before* the regular dependency resolution in Makefile.dep.
+# Hence, an application requiring to provide a feature via a module must do
+# so early on
+
+ifneq (,$(filter soft_spi_as_periph_spi,$(USEMODULE)))
+  FEATURES_PROVIDED += periph_spi
+endif
```

Comment:
This file is new and will need some discussion. I'd like to summon @fjmolinas regarding this.

As a data point: KConfig modelling was just find without crude hacks like this.

Reply:
So far all `FEATURES` are declarative, and this involves having them not be declarative (which is for me the major change), `MODULE` having `FEATURES` I don't see a problem with at a first glance.

What I wonder is if this couldn't be modeled differently to avoid this change. Technically all you need to provide a soft_spi is `periph_timer`, so this could be kept declarative by changing the model to:

```
FEATURES_PROVIDED += periph_spi
FEATURES_PROVIDED += periph_spi_sw
```

All platforms providing a `periph_timer` would provide the softspi feature. And then all modules requiring it would instead do:

```
FEATURES_REQUIRED_ANY += periph_spi|periph_spi_sw
FEATURES_CONFLICT += periph_spi|periph_spi_sw
```

Selecting the right modules to implement that feature becomes a dependency resolution issue.

You would still be able to have `FEATURES_REQUIRED += periph_spi` and `USEMODULE += soft_spi`.  Your `soft_spi_as_periph_spi` is selected by the `periph_spi_sw` feature.

What I like of this is that it keeps it declarative. But I'm not sure how to do it in Kconfig. The `FEATURES_REQUIRED_ANY` and `FEATURES_CONFLICT` are not really modelled. But it seems that a choice with some indirection would work (summon @leandrolanzieri and @MrKevinWeiss as well)

```
choice PERIPH_SPI_API
  bool "periph_spi api backend"
  config MODULE_PERIPH_SPI
      depends on HAS_PERIPH_SPI
  config MODULE_SOFT_SPI_AS_PERIPH_SPI
      depends on HAS_PERIPH_SPI_SW
      select MODULE_SOFT_SPI
endchoice

config MODULE_SOFT_SPI
    bool "Software-implemented SPI"
    depends on HAS_PERIPH_GPIO
    depends on TEST_KCONFIG
    select MODULE_PERIPH_GPIO
    select MODULE_ZTIMER
    select ZTIMER_USEC
```
Note: all used names are place holders.

Reply:
So far all `FEATURES` are declarative, and this involves having them not be declarative (which is for me the major change), `MODULE` having `FEATURES` I don't see a problem with at a first glance.

What I wonder is if this couldn't be modeled differently to avoid this change. Technically all you need to provide a soft_spi is `periph_timer`, so this could be kept declarative by changing the model to:

```
FEATURES_PROVIDED += periph_spi
FEATURES_PROVIDED += periph_spi_sw
```

All platforms providing a `periph_timer` would provide the softspi feature. And then all modules requiring it would instead do:

```
FEATURES_REQUIRED_ANY += periph_spi|periph_spi_sw
FEATURES_CONFLICT += periph_spi|periph_spi_sw
```

Selecting the right modules to implement that feature becomes a dependency resolution issue.

You would still be able to have `FEATURES_REQUIRED += periph_spi` and `USEMODULE += soft_spi`.  Your `soft_spi_as_periph_spi` is selected by the `periph_spi_sw` feature.

What I like of this is that it keeps it declarative. But I'm not sure how to do it in Kconfig. The `FEATURES_REQUIRED_ANY` and `FEATURES_CONFLICT` are not really modelled. But it seems that a choice with some indirection would work (summon @leandrolanzieri and @MrKevinWeiss as well)

```
choice PERIPH_SPI_API
  bool "periph_spi api backend"
  config MODULE_PERIPH_SPI
      depends on HAS_PERIPH_SPI
  config MODULE_SOFT_SPI_AS_PERIPH_SPI
      depends on HAS_PERIPH_SPI_SW
      select MODULE_SOFT_SPI
endchoice

config MODULE_SOFT_SPI
    bool "Software-implemented SPI"
    depends on HAS_PERIPH_GPIO
    depends on TEST_KCONFIG
    select MODULE_PERIPH_GPIO
    select MODULE_ZTIMER
    select ZTIMER_USEC
```
Note: all used names are place holders.

Reply:
Having different providers would not be an issue if we used strictly `depends on` instead of `select`ing some modules (although it comes with the known cumbersome configuration). Based on the @fjmolinas proposal, I think a choice is the best option. I'd add a switch symbol and a feature to be replaced across all the current users of `HAS_PERIPH_SPI` and `MODULE_PERIPH_SPI`.

```
# Kconfig.spi

menuconfig SPI_SUPORT
    bool "SPI support"
    depends on TEST_KCONFIG
    depends on HAS_SPI

choice SPI_IMPLEMENTATION
  bool "SPI implementation"
  depends on SPI_SUPORT

config MODULE_PERIPH_SPI
    bool "Peripheral"
    depends on HAS_SPI

endchoice

# related features
config HAS_SPI
    bool

config HAS_PERIPH_SPI
    bool
    select HAS_SPI
```

```
# drivers/soft_spi/Kconfig

# extend the choice options
choice SPI_IMPLEMENTATION
    depends on SPI_SUPPORT

config MODULE_SOFT_SPI
    bool "Software-implemented SPI"
    depends on HAS_SPI_SW
    select MODULE_PERIPH_GPIO
    select MODULE_ZTIMER
    select ZTIMER_USEC

endchoice

config HAS_SPI_SW
    bool
    depends on HAS_PERIPH_TIMER
    depends on HAS_PERIPH_GPIO
    select HAS_SPI
    default y
```

Reply:
> `MODULE` having `FEATURES` I don't see a problem with at a first glance.

I'm guessing this at least would change the way we produce `info-features-provided`, right? As a dependency resolution would be required now to actually get all the features. Moreover, order of inclusion comes into play now, as we always consider that features are set in stone before requiring/checking them in `Makefile.dep`.

Reply:
I think the right approach would be to let the drivers depend on an `spi` module, rather than on a `periph_spi` feature. The driver likely cares little how the API is provided, as long as the API is there and actually drives an SPI bus. And will also be easier to maintain the list of optional feature dependencies in the single place of the `spi` module, rather than for each driver.

In fact, making this `spi` module a dedicated layer would allow mixing implementations. If an application can keep all hw buses busy via DMA, there could be value in connecting another device via bit-banged SPI. Or, maybe more realistically, some MCUs like the RP2040 have hardware that can be turned into an additional SPI bus (such as the PIO state machines) but doesn't naturally map to the same driver. A high level `spi` module could multiplex all used SPI providers. @benpicco also on this band wagon, as far as I know.

This PR is IMO just a poor waiting to be replaced by a proper solution (: But it might still be useful when porting RIOT to a new MCU until a a hardware SPI driver materialises.